app/notifications/validators.py

```diff
@@ -99,17 +103,72 @@ def check_service_over_daily_message_limit(key_type: ApiKeyType, service: Servic
-        fragments_sent = fetch_daily_sms_fragment_count(service.id)
-        warn_about_daily_sms_limit(service, fragments_sent)
+def check_sms_daily_limit(service: Service, requested_sms=0):
+    messages_sent = fetch_todays_requested_sms_count(service.id)
+    over_sms_daily_limit = (messages_sent + requested_sms) >= service.sms_daily_limit
+
+    # Send a warning when reaching the daily message limit
+    if not over_sms_daily_limit:
+        return
+
+    current_app.logger.info(
+        f"service {service.id} has been rate limited for daily sms use sent {int(messages_sent)} limit {service.sms_daily_limit}"
+    )
+    if service.restricted:
+        raise TrialServiceTooManySMSRequestsError(service.sms_daily_limit)
+    else:
+        raise LiveServiceTooManySMSRequestsError(service.sms_daily_limit)
+
+
+def send_warning_sms_limit_emails_if_needed(service: Service):
+    todays_requested_sms = fetch_todays_requested_sms_count(service.id)
+    nearing_sms_daily_limit = todays_requested_sms >= NEAR_DAILY_LIMIT_PERCENTAGE * service.sms_daily_limit
+    over_sms_daily_limit = todays_requested_sms >= service.sms_daily_limit
+    current_time = datetime.utcnow().isoformat()
+    cache_expiration = int(time_until_end_of_day().total_seconds())
 
+    # Send a warning when reaching 80% of the daily limit
+    if nearing_sms_daily_limit:
+        cache_key = near_sms_daily_limit_cache_key(service.id)
+        if not redis_store.get(cache_key):
+            send_near_sms_limit_email(service)
+            redis_store.set(cache_key, current_time, ex=cache_expiration)
+
+    # Send a warning when reaching the daily message limit
+    if over_sms_daily_limit:
+        cache_key = over_sms_daily_limit_cache_key(service.id)
+        if not redis_store.get(cache_key):
+            send_sms_limit_reached_email(service)
+            redis_store.set(cache_key, current_time, ex=cache_expiration)
 
-def check_rate_limiting(service: Service, api_key: ApiKey, template_type: TemplateType):
+
+def time_until_end_of_day() -> timedelta:
+    """
+    Get timedelta until end of day on the datetime passed, or current time.
+    """
+    dt = datetime.now()
+    tomorrow = dt + timedelta(days=1)
+    return datetime.combine(tomorrow, time.min) - dt
+
+
+def check_sms_limit_increment_redis_send_warnings_if_needed(
+    service: Service, requested_sms=0, key_type: ApiKeyType = KEY_TYPE_NORMAL
+) -> None:
+    if not current_app.config["FF_SPIKE_SMS_DAILY_LIMIT"]:
+        return
+    if key_type == KEY_TYPE_TEST:
+        return
+    if not current_app.config["REDIS_ENABLED"]:
+        return
+
+    check_sms_daily_limit(service, requested_sms)
+    increment_todays_requested_sms_count(service.id, requested_sms)
+    send_warning_sms_limit_emails_if_needed(service)
+
+
+def check_rate_limiting(service: Service, api_key: ApiKey):
     check_service_over_api_rate_limit(service, api_key)
     check_service_over_daily_message_limit(api_key.key_type, service)
```

Comment:
My bad, this is the overall message limit check, which is still needed.  

I think we should get it out of the `check_rate_limiting` function though as that is a bit confusing since the overall daily limit is different than rate limiting. 🤔

Reply:
Yeah I agree that "rate limiting" is not a good name for this function. Let's fix that in a separate PR though, since it's not directly related to the SMS limits and this PR is already getting big 😅